### PR TITLE
[WIP] Compression codecs module

### PR DESCRIPTION
**Pull request by Sindbag**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Implementation of codec pipeline functionality and integration to MergeTree, Log and TinyLog engines.

Design (proposed and approved by @alexey-milovidov ) : 
_ICompressionCodec_ represents interface for **CompressionPipeline** and **CompressionCodec**.
CompressionPipeline describes a sequence of codecs for compression and decompression.
CompressionCodec is one separate compressor or filter, like Delta, LZ4, ZSTD.
Codecs may have arguments, that then should be stored in header part of the compressed block.
For specification of further codecs in header block special continuation bit is used (0x01), that is added to codecs' bytecodes. So the bytecodes section format is:
`[ frist codec bytecode + continuation bit ; first codec arguments ; second codec bytecode, etc... ]`

All internal modifications' sizes also stored in header and could be verified by compressor util. They are used on decompression stage. They stored in reverse to codecs order, and the last size also sums with header size, so the format is:
`[ compressed size = codec N + header sizes ; codec N-1 size ; ... ; original size ]`

CompressionPipeline is specified for each row separately with fallback to default implementation if none specified. Syntax:
`column_name _column_type_ CODEC(LZ4(1), Delta) [ default expression ]`

All codecs implement functions: compress/decompress, parseHeader/writeHeader; compressed sizes methods: getMaxCompressedSize, getCompressedSize, getDecompressedSize (last two used in pipeline).

_CompressionCodecFactory_ is implemented to store and retrieve codecs by names or bytecodes (encoded in header part of blocks). CompressionPipeline also provides _get_pipe_ static methods for pipe creation from AST tree structures, string or ReadBuffer.

LZ4, LZ4HC has new format with argument stored in header, whereas LZ4Old implements old behavior (only bytecode written to block).

Delta codec accepts two parameters: size of element, element type ( 0 - byte, 1 - int, 2 - uint, 3 - float ),
by default the values are calculated from column's data type.

ZSTD also stores argument in header.

Current state:
- [x] - keyword CODEC()
- [x] - AST codec parser
- [x] - Compression settings saver in metadata
- [x] - ZSTD codec
- [x] - ZSTDOld
- [x] - LZ4, LZ4HC codecs
- [x] - LZ4Old codec
- [x] - None codec
- [x] - Delta codec
- [ ] - XORDelta codec
- [ ] - Shuffle codec
- [x] - CompressionPipeline **compress** functionality
- [x] - CompressionPipeline **decompress** functionality
- [x] - _Decompress_ integration in _CompressedReadBufferBase_
- [x] - _Compress_ integration through CompressionSettings in _CompressedWriteBuffer_
- [x] - MergeTree integration
- [x] - Log integration
- [x] - TinyLog integration
- [x] - clickhouse-compressor update
- [ ] - Unit testing
- [x] - Functional testing
- [x] - Performance testing
- [ ] - documentation update

### CODE REVIEW

File: dbms/src/Compression/CompressionCodecDelta.cpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        default:
+            memcpy(dest, source, inputSize);
+            break;
+    }
+    return inputSize;
+}
+
+void CompressionCodecDelta::decompress_bytes(char* source, char* dest, size_t size) const
+{
+    char* tmp_arr = new char[element_size] {};
```
[the remark this answers is not in the document]
Comment (ludv1x) [reply dbms/src/Compression/CompressionCodecDelta.cpp]: The problem is even worse. Here is a memory leak.

File: dbms/src/Compression/CompressionPipeline.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    PODArray<char> buffer1;
+    PODArray<char> buffer2;
+public:
+    ASTPtr codec_ptr;
+    CompressionPipeline(ReadBuffer* header);
+    CompressionPipeline(Codecs& _codecs)
+        : codecs (_codecs)
+    {}
+
+    static CompressionPipePtr get_pipe(ReadBuffer* header);
```
Comment (ludv1x) [inline dbms/src/Compression/CompressionPipeline.h]: Bad name `createPipeline` is better.

File: dbms/src/Compression/CompressionPipeline.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+using CompressionCodecPtr = std::shared_ptr<ICompressionCodec>;
+using Codecs = std::vector<CompressionCodecPtr>;
+
+class CompressionPipeline;
+using CompressionPipePtr = std::shared_ptr<CompressionPipeline>;
```
Comment (ludv1x) [inline dbms/src/Compression/CompressionPipeline.h]: Bad abbreviation, `CompressionPipelinePtr` is better.

**Comment by zhang2014** [comment]
@ludv1x @Sindbag  Is there any new progress about this feature you can share with us? We really need this :)

**Comment by alexey-milovidov** [comment]
@alesapin will take this task.

**Comment by zhang2014** [comment]
Maybe this can be left to me? I have opened another pull request #2586 : )

**Comment by alexey-milovidov** [comment]
We can pause this task till the end of September in favor of your changes.

**Comment by zhang2014** [comment]
OK, thanks. I will finish it as soon as possible.

**Comment by alexey-milovidov** [comment]
Superseded by #2586

**Pull request closed without merging**